### DISCUSSION TITLE
Only able to switch PC on with home button, only able to switch off with homebridge command

### DISCUSSION THREAD

**Discussion by safetrvls** [AlexGustafsson/homebridge-wol#180] [Questions]
I'm new to using terminal and linux so please forgive any basic misunderstandings!

I've followed the guide on the wiki as closely as possible and I've ended up being able to use the homekit button to switch on my Windows 10 PC, but not switch it off. Is there supposed to be a second button in the home app that triggers the shutdown process or can the PC button function as a true on/off switch?

The exact opposite is happening on the homebridge terminal, I can switch the PC off with the suggested command, 
"ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa ******@********* shutdown /s"

But replacing "shutdown" with "start" does not switch the PC on. Clearly the functionality is there if I can do both, but what am I missing?

This is my current config:

name": "PC",
            "ip": "***********",
            "pingInterval": 2,
            "pingsToChange": 5,
            "pingTimeout": 1,
            "pingCommand": "ping",
            "pingCommandTimeout": 0,
            "mac": "************",
            "startCommand": "start",
            "startCommandTimeout": 0,
            "wakeGraceTime": 45,
            "wakeCommand": "wake",
            "wakeCommandTimeout": 0,
            "shutdownCommand": "shutdown",
            "shutdownGraceTime": 60,
            "shutdownCommandTimeout": 0,
            "log": true,
            "logPinger": true,
            "debugLog": true,
            "returnEarly": false,
            "accessory": "NetworkDevice"
            
Thanks for building this awesome plugin, any help would be much appreciated!

**Comment by AlexGustafsson** [comment]
Your start command and shutdown command looks off. From the README:

> [startCommand] Command to run in order to start the machine
> [shutdownCommand] Command to run in order to shut down the remote machine

The start command is only necessary if you want to start your device with something other than WoL. To use WoL as intended, you can remove `startCommand` and make sure that `mac` is correctly set.

The shutdown command is necessary if you wish to shut down the device as there's no functionality like WoL that can be used. The command you've mentioned, `ssh -o ...` is the value you should use for the `shutdownCommand`.

Basically your config should look a bit like this:

```json
{
  "accessory": "NetworkDevice",
  "name": "PC",
  "ip": "192.168.1.112",
  "mac": "xxxx",
  "shutdownCommand": "ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @*** shutdown /s"
}
```

The intended functionality is to expose a switch in the Home app that, when turned on, turns on your computer and when turned off, turns your computer off.

**Reply by safetrvls** [reply]
Thanks very much for the reply, I tried implementing it just as you said but shutdown still doesn't work from the accessories page of homebridge or the home app.
 
I tried pasting the config you posted exactly as it is written with the PC's IP and Mac address put in, as well as the IP address in the shutdown command, and it didn't work. I also tried a more verbose version than keeps the default values for timeouts etc, also gives me no response. Could some setting in windows be wrong?
 
 "name": "PC",
            "ip": "[PC IP address]",
            "pingInterval": 2,
            "pingsToChange": 5,
            "pingTimeout": 1,
            "pingCommand": "ping",
            "pingCommandTimeout": 0,
            "mac": "[PC MAC address]",
            "startCommandTimeout": 0,
            "wakeGraceTime": 45,
            "wakeCommand": "wake",
            "wakeCommandTimeout": 0,
            "shutdownCommand": "ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @[PC IP address] shutdown /s",
            "shutdownGraceTime": 60,
            "shutdownCommandTimeout": 0,
            "log": true,
            "logPinger": true,
            "debugLog": true,
            "returnEarly": false,
            "accessory": "NetworkDevice"

Start up still works perfectly. Thanks for bearing with me!

**Reply by AlexGustafsson** [reply]
Could you post the logs? I assume you'll find logs such as `Error: Command failed: ssh`. This would indicate that the shutdown command is run correctly, but that it fails. It may fail for any number of reasons. It could be that home bridge cannot access the key in `/homebridge/id_rsa`, that the IP address is bad, that SSH fails altogether and so on.

**Reply by safetrvls** [reply]
Here is the relevant log for the shutdown issue. You're right that something is wrong with the ssh, but I'm not command line literate enough to really understand... Does this tell you where the issue is?

[PC] Attempting to shut down "PC" (192.168.1.233) using "ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @192.168.1.xxx shutdown /s"
[11/21/2021, 7:23:29 PM] [PC] An error occured while trying to shut down "PC" (192.168.1.xxx): Error: Command failed: ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @192.168.1.233 shutdown /s
/bin/sh: 1: ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @192.168.1.xxx shutdown /s: not found

    at ChildProcess.exithandler (child_process.js:383:12)
    at ChildProcess.emit (events.js:400:28)
    at maybeClose (internal/child_process.js:1058:16)
    at Socket.<anonymous> (internal/child_process.js:443:11)
    at Socket.emit (events.js:400:28)
    at Pipe.<anonymous> (net.js:686:12) {
  killed: false,
  code: 127,
  signal: null,
  cmd: 'ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @192.168.1.xxx shutdown /s',
  stdout: '',
  stderr: '/bin/sh: 1: ssh -o StrictHostKeyChecking=no -i /homebridge/id_rsa @192.168.1.xxx shutdown /s: not found\n'
}
[11/21/2021, 7:23:29 PM] [PC] Waiting for shutdown grace time (15000ms) for "PC" (192.168.1.xxx)

**Reply by AlexGustafsson** [reply]
This seems to be another instance of #156, so it's a bug. What version are you running of the plugin and of NodeJS? This issue has only popped up twice for over a year since the last stable version, which is weird. It might be platform specific, but it's something we should fix.

**Reply by AlexGustafsson** [reply]
Also, what platform do you run home bridge on? Windows or Linux?

**Reply by safetrvls** [reply]
Interesting! I'm running homebridge on raspbian. NodeJs v14.18.1 and homebridge-wol v4.3.0

**Reply by AlexGustafsson** [reply]
Could you try to run the latest version? See #124. Perhaps the issue is already fixed.

**Comment by safetrvls** [comment]
Would love to, but I'm not sure how... Tried running npm install homebridge-wol@beta in the homebridge terminal and it said

  npm WARN saveError ENOENT: no such file or directory, open '/var/lib/homebridge/package.json'
  npm WARN enoent ENOENT: no such file or directory, open '/var/lib/homebridge/package.json'
  npm WARN homebridge No description
  npm WARN homebridge No repository field.
  npm WARN homebridge No README data
  npm WARN homebridge No license field.
  
  + homebridge-wol@5.0.0-beta.1
  updated 1 package and audited 11 packages in 1.247s
  found 0 vulnerabilities

Version number is still 4.3.0. I know I'm lacking some very basic knowledge here, sorry.

**Reply by AlexGustafsson** [reply]
You're nearly there. Try to install the package globally.

```shell
npm install --global homebridge-wol@beta 
```

**Reply by safetrvls** [reply]
Thanks, on 5.0.0-beta.1 now. Unfortunately now it won't switch on through the plugin either. Maybe this is just not for me...

EDIT: updating the version made the wol settings disappear from the homebridge config even though the settings were still in the gui. Renaming the device and saving it restored the settings in the config and I'm back to being able to wake but not shutdown.
